### PR TITLE
Daniel reorder

### PR DESCRIPTION
**Pull request by dkrasner**
### Main Points ###

This PR primarily introduces re-ordering, but in doing creates a new message type (and related subscribers), refactors how views are added, imposes an order sync between subparts and subpart-views, amongst other changes. 

#### Move Commands ####

[Move commands](https://github.com/dkrasner/Simpletalk/compare/daniel-reorder?expand=1#diff-efd61be9944474fec7ba0b89d7a931fedae606964b569ff27b2766271c5244b9R91) have been added to the part class. 

The first thing to note is that these correspond to the list subpart order, so `moveToLast` = move to the last in the list; `moveDown` = move one down in the list. This means that in the view `moveToLast` move a card to the last one in the order, but also move a part view on a card **overlaying** any of its siblings. This seems reasonably intuitive, since we have an order as to how things are added and also cards are in order and corresponds to the `number` property. We could also add commands like `moveAbove` or `moveBelow`  tbd

The second thing to note is that these are overridden for `Stacks` and `Cards`, since we insist that they come first as children in their respective parents. 

So now we have different parts deciding how they should be ordered. Once decided they tell their owners to re-order, [like so](https://github.com/dkrasner/Simpletalk/compare/daniel-reorder?expand=1#diff-efd61be9944474fec7ba0b89d7a931fedae606964b569ff27b2766271c5244b9R539) and this [triggers](https://github.com/dkrasner/Simpletalk/compare/daniel-reorder?expand=1#diff-efd61be9944474fec7ba0b89d7a931fedae606964b569ff27b2766271c5244b9R602) `viewChanged` type message (in particular `changeType: subpart-order`). 

#### view changes messages ####

Like property changes view changes have subscribers and change handlers. For the moment we have two types of changes:

[supart-order](https://github.com/dkrasner/Simpletalk/compare/daniel-reorder?expand=1#diff-efd61be9944474fec7ba0b89d7a931fedae606964b569ff27b2766271c5244b9R602) which sorts out how to insert the view into the proper place (either by appending child or inserting node)

and 

[subpart-new](https://github.com/dkrasner/Simpletalk/compare/daniel-reorder?expand=1#diff-f8a644f39713f7d56b51b944769a8733fceb081884704c2a73b93c232b554560R362) which in general just appends the new view. The reason for this change is that previously we had no clean or centralized ways of adding views. `System.newView` would deal with checking what type of view it was adding (card, stack, window) etc, making it a bit of mess; de-serialization had its own adding views. This introduced bugs like this one #40. Having views, like parts, deal with how their add their respective children seems sound. 

#### Examples ####
I have added a `reorder-example.html` to play with. 

#### Tests ####

I will write more tests before merging

#### Known issues ####
I am still finding the de-serialized examples don't respect subpart and child order. I don't really see at the moment why this is so: subparts are added in the same order as the subpartViews  but it needs to be sorted out before merge. Maybe @darth-cheney you know...

Keeping the nav open and re-ordering cards (like in the example) can break it as in this issue #6

### CODE REVIEW

**Comment by dkrasner** [comment]
@darth-cheney I've decided to re-work some of the design here and now have parts be responsible for handling how their subparts are re-ordered. Look  [here](https://github.com/dkrasner/Simpletalk/pull/41/commits/1842c14c61831b92b934056555e2094e0b9769a8#diff-efd61be9944474fec7ba0b89d7a931fedae606964b569ff27b2766271c5244b9R96)

the reasons for this are the following:

semantically it makes sense to have move* commands are private handlers on the parts themselves, i.e. `tell card 2 to moveUp` is pretty natural. It would be less natural to refer to subparts somehow in a statements like that. 

on the other hand a stack knows how to add its subparts better than the parts. For example, if it's adding a card it will put it in the card block which precedes all other parts, if it's adding a non-card it will put it after the card block, etc. Similar with world. 

So now a command like `moveUp` on a part will simply call `part._owner.moveSubpartUp(this)`. This also gives a nice interface for moving subparts around, i.e. just tell the part which to move and how. 

I've added a lot of [tests](https://github.com/dkrasner/Simpletalk/blob/daniel-reorder/js/objects/tests/test-private-commands-with-preload.js) to underline how things are supposed move around.

**Comment by darth-cheney** [comment]
@dkrasner I have three failing tests on the working branch related to movement. Here are the errors:
```
  1) Private command tests
       Subpart re-order with cards
         MoveToFirst (cannot move non-card part before cards):
     TypeError: Failed to execute 'insertBefore' on 'Node': parameter 1 is not of type 'Node'.
      at Object.exports.convert (node_modules/jsdom/lib/jsdom/living/generated/Node.js:25:9)
      at HTMLElement.insertBefore (node_modules/jsdom/lib/jsdom/living/generated/Node.js:356:26)
      at HTMLElement.subpartOrderChanged (js/objects/views/navigator/CardRow.js:155:18)
      at HTMLElement.primHandleViewChange (js/objects/views/PartView.js:328:16)
      at HTMLElement.receiveMessage (js/objects/views/PartView.js:284:18)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Stack.sendMessage (js/objects/parts/Part.js:397:30)
      at /home/ecgade/projects/simpletalk-06-02/SimpleTalk/js/objects/parts/Part.js:615:18
      at Set.forEach (<anonymous>)
      at Stack.viewChanged (js/objects/parts/Part.js:614:31)
      at Stack.subpartOrderChanged (js/objects/parts/Part.js:622:14)
      at Stack.moveSubpartToFirst (js/objects/parts/Stack.js:254:18)
      at Button.<anonymous> (js/objects/parts/Part.js:98:73)
      at Button.receiveCmd (js/objects/parts/Part.js:457:26)
      at Button.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Button.sendMessage (js/objects/parts/Part.js:397:30)
      at Object._e20‍.r.System._commandHandlers.tell (js/objects/System.js:1015:16)
      at Object.receiveCommand (js/objects/System.js:296:26)
      at Object.receiveMessage (js/objects/System.js:210:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at WorldStack.sendMessage (js/objects/parts/Part.js:397:30)
      at WorldStack.delegateMessage (js/objects/parts/WorldStack.js:230:21)
      at WorldStack.receiveCmd (js/objects/parts/Part.js:471:21)
      at WorldStack.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Stack.sendMessage (js/objects/parts/Part.js:397:30)
      at Stack.delegateMessage (js/objects/parts/Part.js:390:21)
      at Stack.receiveCmd (js/objects/parts/Part.js:471:21)
      at Stack.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Card.sendMessage (js/objects/parts/Part.js:397:30)
      at Card.delegateMessage (js/objects/parts/Part.js:390:21)
      at Card.receiveCmd (js/objects/parts/Part.js:471:21)
      at Card.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Card.sendMessage (js/objects/parts/Part.js:397:30)
      at Context.<anonymous> (js/objects/tests/test-private-commands-with-preload.js:438:25)
      at processImmediate (node:internal/timers:464:21)

  2) Private command tests
       Subpart re-order with cards
         MoveUp (can move non-card parts up before non-card parts):
     TypeError: Failed to execute 'insertBefore' on 'Node': parameter 1 is not of type 'Node'.
      at Object.exports.convert (node_modules/jsdom/lib/jsdom/living/generated/Node.js:25:9)
      at HTMLElement.insertBefore (node_modules/jsdom/lib/jsdom/living/generated/Node.js:356:26)
      at HTMLElement.subpartOrderChanged (js/objects/views/navigator/CardRow.js:155:18)
      at HTMLElement.primHandleViewChange (js/objects/views/PartView.js:328:16)
      at HTMLElement.receiveMessage (js/objects/views/PartView.js:284:18)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Stack.sendMessage (js/objects/parts/Part.js:397:30)
      at /home/ecgade/projects/simpletalk-06-02/SimpleTalk/js/objects/parts/Part.js:615:18
      at Set.forEach (<anonymous>)
      at Stack.viewChanged (js/objects/parts/Part.js:614:31)
      at Stack.subpartOrderChanged (js/objects/parts/Part.js:622:14)
      at Stack.moveSubpartUp (js/objects/parts/Stack.js:240:18)
      at Button.<anonymous> (js/objects/parts/Part.js:96:68)
      at Button.receiveCmd (js/objects/parts/Part.js:457:26)
      at Button.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Button.sendMessage (js/objects/parts/Part.js:397:30)
      at Object._e20‍.r.System._commandHandlers.tell (js/objects/System.js:1015:16)
      at Object.receiveCommand (js/objects/System.js:296:26)
      at Object.receiveMessage (js/objects/System.js:210:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at WorldStack.sendMessage (js/objects/parts/Part.js:397:30)
      at WorldStack.delegateMessage (js/objects/parts/WorldStack.js:230:21)
      at WorldStack.receiveCmd (js/objects/parts/Part.js:471:21)
      at WorldStack.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Stack.sendMessage (js/objects/parts/Part.js:397:30)
      at Stack.delegateMessage (js/objects/parts/Part.js:390:21)
      at Stack.receiveCmd (js/objects/parts/Part.js:471:21)
      at Stack.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Card.sendMessage (js/objects/parts/Part.js:397:30)
      at Card.delegateMessage (js/objects/parts/Part.js:390:21)
      at Card.receiveCmd (js/objects/parts/Part.js:471:21)
      at Card.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Card.sendMessage (js/objects/parts/Part.js:397:30)
      at Context.<anonymous> (js/objects/tests/test-private-commands-with-preload.js:451:25)
      at processImmediate (node:internal/timers:464:21)

  3) Private command tests
       Subpart re-order with cards
         MoveDown (can move non-card parts down after non-card parts):
     TypeError: Failed to execute 'insertBefore' on 'Node': parameter 1 is not of type 'Node'.
      at Object.exports.convert (node_modules/jsdom/lib/jsdom/living/generated/Node.js:25:9)
      at HTMLElement.insertBefore (node_modules/jsdom/lib/jsdom/living/generated/Node.js:356:26)
      at HTMLElement.subpartOrderChanged (js/objects/views/navigator/CardRow.js:155:18)
      at HTMLElement.primHandleViewChange (js/objects/views/PartView.js:328:16)
      at HTMLElement.receiveMessage (js/objects/views/PartView.js:284:18)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Stack.sendMessage (js/objects/parts/Part.js:397:30)
      at /home/ecgade/projects/simpletalk-06-02/SimpleTalk/js/objects/parts/Part.js:615:18
      at Set.forEach (<anonymous>)
      at Stack.viewChanged (js/objects/parts/Part.js:614:31)
      at Stack.subpartOrderChanged (js/objects/parts/Part.js:622:14)
      at Stack.moveSubpartDown (js/objects/parts/Stack.js:212:18)
      at Button.<anonymous> (js/objects/parts/Part.js:97:70)
      at Button.receiveCmd (js/objects/parts/Part.js:457:26)
      at Button.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Button.sendMessage (js/objects/parts/Part.js:397:30)
      at Object._e20‍.r.System._commandHandlers.tell (js/objects/System.js:1015:16)
      at Object.receiveCommand (js/objects/System.js:296:26)
      at Object.receiveMessage (js/objects/System.js:210:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at WorldStack.sendMessage (js/objects/parts/Part.js:397:30)
      at WorldStack.delegateMessage (js/objects/parts/WorldStack.js:230:21)
      at WorldStack.receiveCmd (js/objects/parts/Part.js:471:21)
      at WorldStack.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Stack.sendMessage (js/objects/parts/Part.js:397:30)
      at Stack.delegateMessage (js/objects/parts/Part.js:390:21)
      at Stack.receiveCmd (js/objects/parts/Part.js:471:21)
      at Stack.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Card.sendMessage (js/objects/parts/Part.js:397:30)
      at Card.delegateMessage (js/objects/parts/Part.js:390:21)
      at Card.receiveCmd (js/objects/parts/Part.js:471:21)
      at Card.receiveMessage (js/objects/parts/Part.js:405:29)
      at Object.sendMessage (js/objects/System.js:197:23)
      at Card.sendMessage (js/objects/parts/Part.js:397:30)
      at Context.<anonymous> (js/objects/tests/test-private-commands-with-preload.js:477:25)
      at processImmediate (node:internal/timers:464:21)
```

File: js/objects/System.js [hunk 1]
```diff
@@ -562,27 +556,11 @@ const System = {
...
-                    let pane = window._shadowRoot.querySelector('.st-window-pane');
-                    pane.append(newView);
-                });
-            }
-            parentElement.appendChild(newView);
-        }
+        this.sendMessage({
+            type: "viewChanged",
+            changeName: "subpart-new",
+            args: [newView]
```
Comment (darth-cheney) [inline js/objects/System.js:562]: One thing we said early on is that we wanted messages to be serializable. Is that still the case? If so, we need to make the argument passed with this message to be a part id. The element lookup can be handles by any corresponding view that receives the message.

Comment (dkrasner) [reply js/objects/System.js:562]: @darth-cheney you mean?

```
this.sendMessage({
type: "viewChanged",
changeName: "subpart-new",
partId: ID
args: [newView]},
target)?
```

**Comment by dkrasner** [comment]
> @dkrasner I have three failing tests on the working branch related to movement. Here are the errors:

@darth-cheney fixed (nav again..) 


also i pushed the "fix" to the nav shadow DOM issue we had fun with (#6) 

it's not a perfect fix, there are some minor overlay-number issues that can still come up, with the same problem of the css disappearing but it's not a big deal atm

File: js/objects/System.js (hunk 1 above)
Comment (darth-cheney) [reply js/objects/System.js:562]: `newView` is a DOM element here, which is not technically serializable (unless we want to serialize it's HTML). That is the argument that should be an id instead. Then the receiver can lookup the corresponding element based on the id as needed.

**Pull request merged**